Always use Bicleaner AI

eu9ene opened the pull request:
- Always use Bicleaner-AI and deprecate Bicleaner
- If a language pair is not supported use the multilingual model
- Always use full models 
- Download models using bicleaner tool
- Rewrite and simplify model pack downloading, replace static URLs with a python script 
- Add tests
- Refactor bicleaner kinds
- Add filtered lines to artifacts: https://firefox-ci-tc.services.mozilla.com/tasks/FZR3B6bATa-DpxMv3PqbDg/runs/0
- Increased maxRunTime for some steps that will work with huge datasets and might perform resource heavy operations

Fixes #55 #360

marco-c commented on pipeline/bicleaner/bicleaner.sh:
```diff
@@ -76,7 +69,7 @@ else
                export CUDA_VISIBLE_ARRAY=(${CUDA_VISIBLE_DEVICES//,/ })
                export CUDA_VISIBLE_DEVICES=${CUDA_VISIBLE_ARRAY[$(($2-1))]}
-               bicleaner-ai-classify --scol ${scol} --tcol ${tcol} - - $1
+               bicleaner-ai-classify --disable_hardrules --scol ${scol} --tcol ${tcol} - - $1
```
Shouldn't `--disable_hardrules` only be used in the case of multilingual bicleaner?

eu9ene replied:
Good catch! I was under impression that we can skip hard rules for everything but based on this [comment](https://github.com/mozilla/firefox-translations-training/issues/71#issuecomment-1040029744) they hurt only multilingual model

eu9ene commented on pipeline/bicleaner/bicleaner.sh:
```diff
@@ -94,6 +87,10 @@ else
     awk -v threshold=${bicleaner_threshold} -F"\t" '{if ($3>threshold) {print $0}}' |
     ${COMPRESSION_CMD} >"${output_prefix}.best.${ARTIFACT_EXT}"
 
+  ${COMPRESSION_CMD} -dc "${output_prefix}.scored.${ARTIFACT_EXT}" |
```
I also added saving to artifacts filtered sentences

marco-c replied:
nit: it'd be nice to do these related but not completely related changes in separate commits or separate PRs, so the changes are easier to review

eu9ene commented on taskcluster/kinds/merge-corpus/kind.yml:
```diff
@@ -13,18 +13,7 @@ transforms:
     - taskgraph.transforms.task:transforms
 
 kind-dependencies:
-    # This kinds handles two different parts of the graph. The first is `merge-corpus`, which
```
those comments looked outdated since we now have only bicleaner ai and a separate kind for merge-devset

eu9ene commented on taskcluster/translations_taskgraph/transforms/find_upstreams.py:
```diff
@@ -63,7 +63,7 @@ def get_cleaning_type(upstreams):
 
     for upstream in upstreams:
-        if upstream.kind not in ("bicleaner", "clean-corpus"):
+        if upstream.kind not in ("clean-corpus", "bicleaner"):
```
no reason here, I thought I would need to remove bicleaner here but it broke things, so I added to back. I do not completely understand how this part works

bhearsum replied:
This is because `bicleaner` here refers to a `kind` (both `bicleaner` and `bicleaner-ai` are in the `bicleaner` kind`). Basically, this tuple needs to contain any kinds that contain cleaning tasks.

marco-c commented on docs/bicleaner.md:
```diff
@@ -5,17 +5,11 @@ parent: Data cleaning
-  * [https://github.com/bitextor/bicleaner-ai](https://github.com/bitextor/bicleaner-ai)
-  * [https://github.com/bitextor/bicleaner](https://github.com/bitextor/bicleaner)
+[Bicleaner AI](https://github.com/bitextor/bicleaner-ai) is a tool that aims at detecting noisy sentence pairs in a parallel corpus. 
+If specialized model for a language pair is not available it will fallback to downloading multilingual en-xx model.
```
```suggestion
If i specialized model for a language pair is not available, it will fallback to downloading a multilingual en-xx model.
```

marco-c commented on pipeline/bicleaner/download_pack.py:
```diff
@@ -0,0 +1,119 @@
+
+
+def _compress_dir(dir_path: str, compression_cmd: str) -> str:
+    print(f"Compressing {dir_path}")
```
Nit: use `logging`

eu9ene replied:
We would need to switch all our scripts to logging then. We can think of it separately. The only benefit logging brings for such simple scripts that we run under Taskcluster is probably log level of the message.

marco-c commented on pipeline/bicleaner/download_pack.py:
```diff
@@ -0,0 +1,119 @@
+def _compress_dir(dir_path: str, compression_cmd: str) -> str:
+    print(f"Compressing {dir_path}")
+    if compression_cmd not in ["gzip", "zstd", "zstdmt", "pigz"]:
+        raise ValueError("Unsupported compression tool.")
```
nit: Add compression_cmd to the exception string, so if it happens, we quickly know what's missing

marco-c commented on pipeline/bicleaner/download_pack.py:
```diff
@@ -0,0 +1,119 @@
+    if compression_cmd not in ["gzip", "zstd", "zstdmt", "pigz"]:
+        raise ValueError("Unsupported compression tool.")
+
+    tarball_path = dir_path + ".tar"
```
nit:
```suggestion
    tarball_path = f"{dir_path}.tar"
```

marco-c replied:
nit: Here and in a few other places, string concatenation with "+" is not very Pythonic

marco-c commented on pipeline/bicleaner/download_pack.py:
```diff
@@ -0,0 +1,119 @@
+
+        result.check_returncode()
+
+    # Compress downloaded pack and
```
The comment is cut off here

marco-c commented on pipeline/bicleaner/download_pack.py:
```diff
@@ -0,0 +1,119 @@
+    print("Done")
+
+
+def main(args: Optional[list[str]] = None) -> None:
```
nit: instead of `Optional`, you can do `args: list[str] | None`

eu9ene replied:
The support for this was added in 3.10 (https://peps.python.org/pep-0604/). We might use earlier version in docker images. Also I prefer the code to be compatible at least with 3.9. There's not much difference here.

marco-c commented on taskcluster/kinds/merge-devset/kind.yml:
```diff
@@ -12,18 +12,10 @@ transforms:
     - translations_taskgraph.transforms.cached_tasks:transforms
     - taskgraph.transforms.task:transforms
 
+# todo: why bicleaner and clean-corpus are required here? it doesn't work without them
```
What is the issue?
I bet @bhearsum or @gabrielBusta can help

bhearsum replied:
I'd have to know more about what doesn't work to say anything with certainty, but in general you need to have `kind-dependencies` set if any tasks in this kind are generated from another kind.

(I am a little bit surprised that `clean-corpus` is required here....that looks like it may be an historical artifact from when all of the `merge` tasks were in the same kind...)

eu9ene replied:
merging devest does not depend on cleaning or bicleaner. We merge raw datasets here. When I remove those two and validate the taskgraph it says:
```
2024-01-22 15:52:17,682 - ERROR - Error loading tasks for kind merge-devset:
Traceback (most recent call last):
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 311, in _run
    new_tasks = kind.load_tasks(
                ^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 76, in load_tasks
    tasks = [
            ^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 76, in <listcomp>
    tasks = [
            ^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1347, in check_run_task_caches
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1276, in check_task_dependencies
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1262, in check_task_identifiers
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1243, in chain_of_trust
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1236, in add_github_checks
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1061, in build_task
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1029, in add_index_routes
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 957, in process_treeherder_metadata
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 919, in validate
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 907, in task_name_from_label
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 863, in set_defaults
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 842, in set_implementation
    for task in tasks:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/cached_tasks.py", line 117, in cache_task
    for task in order_tasks(config, tasks):
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/cached_tasks.py", line 22, in order_tasks
    pending = deque(tasks)
              ^^^^^^^^^^^^
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/cached_tasks.py", line 69, in add_cache
    for job in jobs:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/cached_tasks.py", line 55, in resolved_keyed_by_fields
    for job in jobs:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/base.py", line 144, in __call__
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 361, in make_task_description
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 236, in use_fetches
    for task in order_tasks(config, tasks):
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/cached_tasks.py", line 22, in order_tasks
    pending = deque(tasks)
              ^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 164, in add_resource_monitor
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 152, in set_label
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 136, in set_implementation
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 116, in rewrite_when_to_optimization
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/base.py", line 144, in __call__
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task_context.py", line 85, in render_task
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/base.py", line 144, in __call__
    for task in tasks:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/find_upstreams.py", line 97, in upstreams_for_locales
    for job in jobs:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/find_upstreams.py", line 82, in resolve_keyed_by_fields
    cleaning_type = get_cleaning_type(config.kind_dependencies_tasks.values())
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/find_upstreams.py", line 75, in get_cleaning_type
    raise Exception("Unable to find cleaning type!")
Exception: Unable to find cleaning type!
Traceback (most recent call last):
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/main.py", line 907, in main
    return args.command(vars(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/main.py", line 430, in show_taskgraph
    ret = generate_taskgraph(options, parameters, logdir)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/main.py", line 192, in generate_taskgraph
    out = format_taskgraph(options, spec, logfile(spec))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/main.py", line 148, in format_taskgraph
    tg = getattr(tgg, options["graph_attr"])
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 170, in full_task_graph
    return self._run_until("full_task_graph")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 425, in _run_until
    k, v = next(self._run)
           ^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 311, in _run
    new_tasks = kind.load_tasks(
                ^^^^^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 76, in load_tasks
    tasks = [
            ^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/generator.py", line 76, in <listcomp>
    tasks = [
            ^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1347, in check_run_task_caches
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1276, in check_task_dependencies
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1262, in check_task_identifiers
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1243, in chain_of_trust
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1236, in add_github_checks
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1061, in build_task
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 1029, in add_index_routes
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 957, in process_treeherder_metadata
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 919, in validate
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 907, in task_name_from_label
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 863, in set_defaults
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task.py", line 842, in set_implementation
    for task in tasks:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/cached_tasks.py", line 117, in cache_task
    for task in order_tasks(config, tasks):
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/cached_tasks.py", line 22, in order_tasks
    pending = deque(tasks)
              ^^^^^^^^^^^^
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/cached_tasks.py", line 69, in add_cache
    for job in jobs:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/cached_tasks.py", line 55, in resolved_keyed_by_fields
    for job in jobs:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/base.py", line 144, in __call__
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 361, in make_task_description
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 236, in use_fetches
    for task in order_tasks(config, tasks):
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/cached_tasks.py", line 22, in order_tasks
    pending = deque(tasks)
              ^^^^^^^^^^^^
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 164, in add_resource_monitor
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 152, in set_label
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 136, in set_implementation
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/run/__init__.py", line 116, in rewrite_when_to_optimization
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/base.py", line 144, in __call__
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/task_context.py", line 85, in render_task
    for task in tasks:
  File "/Users/epavlov/Library/Caches/pypoetry/virtualenvs/firefox-translations-training-U0HVfMmK-py3.11/lib/python3.11/site-packages/taskgraph/transforms/base.py", line 144, in __call__
    for task in tasks:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/find_upstreams.py", line 97, in upstreams_for_locales
    for job in jobs:
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/find_upstreams.py", line 82, in resolve_keyed_by_fields
    cleaning_type = get_cleaning_type(config.kind_dependencies_tasks.values())
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epavlov/mozilla/source/origin/firefox-translations-training/taskcluster/translations_taskgraph/transforms/find_upstreams.py", line 75, in get_cleaning_type
    raise Exception("Unable to find cleaning type!")
Exception: Unable to find cleaning type!

```

eu9ene replied:
this issue is not related to this PR though

bhearsum replied:
Ah, I see. Yep - that's an oversight from when `merge-devset` moved to its own kind. https://github.com/mozilla/firefox-translations-training/pull/385 fixes it.

marco-c commented on taskcluster/translations_taskgraph/transforms/find_upstreams.py:
```diff
@@ -63,7 +63,7 @@ def get_cleaning_type(upstreams):
 
     for upstream in upstreams:
-        if upstream.kind not in ("bicleaner", "clean-corpus"):
+        if upstream.kind not in ("clean-corpus", "bicleaner"):
```
Since you ended up not changing anything, can you put the order back so this doesn't show up in the diff?

marco-c commented:
I see a downside in the approach. Previously we had all models listed. It was a bit painful when you wanted to add a new one, but it allowed us to specify exactly which model version we wanted to download and so it allowed us to make use of caching effectively (because we knew when to invalidate things).
Unless I'm missing something, with the new approach we are not specifying which model to download, so we will always use the latest one, which makes things not easily reproducible.
Also, when downloading a model for a pair like en-ru where we use the multilingual one, how do we invalidate the cache in case a en-ru non-multilingual model becomes available?

If we want to accept the downside in exchange for avoiding to list all models, we should probably document the decision somewhere.

bhearsum commented on taskcluster/kinds/bicleaner-model/kind.yml:
```diff
@@ -0,0 +1,73 @@
+                - attributes
+                - run.command
+        worker-type: b-cpu
+        expires-after: "90 days"
```
note: expiring this after 90 days will almost certainly cause everything downstream of them (like training task) to expire at the same time. Also, this expiry is only partly related to how long something stays cached -- the task expiring is one way a task will no longer be cached, but we have other ways to refresh a task in the cache (eg: by changing something that affects the cache digest).

I'm assuming that the reason you've set this to 90 days is to make sure we don't use outdated models for too long. If that is true, I suggest that we find a way to easily change the cache digest instead. That way we can cache models for as long as possible (eg: for a long lived training branch), and force a refresh through other means. The most straightforward way to do this would be to add a new file on disk to `resources`, which we could change anytime we want to refresh the bicleaner model cache.

eu9ene replied:
Interesting, I didn't know it works that way. I'll remove it.

bhearsum reviewed (changes requested):
I'm marking this as Request changes due to the note below about the bicleaner model cache. Hopefully I'm understanding the intent there - please feel free to ping me to discuss.

todo: With non-ai bicleaner going away, the `bicleaner` in https://github.com/mozilla/firefox-translations-training/blob/71013bcea0e4647d04d508daf45fe2a96c27ef0d/taskcluster/translations_taskgraph/transforms/find_upstreams.py#L71 is now dead code and can be removed.

gregtatum reviewed (commented):
I had a few small pieces of feedback looking through this, but others have done a more thorough review, so I'm just leaving a comment.

eu9ene commented:
> I see a downside in the approach. Previously we had all models listed. It was a bit painful when you wanted to add a new one, but it allowed us to specify exactly which model version we wanted to download and so it allowed us to make use of caching effectively (because we knew when to invalidate things). Unless I'm missing something, with the new approach we are not specifying which model to download, so we will always use the latest one, which makes things not easily reproducible. Also, when downloading a model for a pair like en-ru where we use the multilingual one, how do we invalidate the cache in case a en-ru non-multilingual model becomes available?
> 
> If we want to accept the downside in exchange for avoiding to list all models, we should probably document the decision somewhere.

The idea was to delegate downloading to bicleaner and easily switch between the full and lite models + hide multilingual one to reduce complexity on Taskcluster side. I looked at the code and you're right, it does download latest which is not ideal, but there was just one release of the models so I don't think this will become a problem for us. Paracrawl project is over and I don't know whether there are any plans to retrain the models or train new languages.  Let's keep an eye on this. 

As for caching, we can just set expiration to this task. And anyway we're likely always interested in the latest models as soon as they are compatible with the code. I assume they'll change the URLs and HF repo names if they ever train new incompatible models.

We download fastText model in the same way, now with OpusCleaner. We also download tons of datasets every time which also might get updated and we don't cache them well in the same way. We can check but I don't think downloading is a huge cost for us compared to GPUs. I prefer flexibility to complexity here.

I'll add docs about this.

marco-c commented:
My concerns are more with reproducibility than cost in this case, but if we don't expect the models to change often and we document this well, it seems OK.

eu9ene commented:
> My concerns are more with reproducibility than cost in this case, but if we don't expect the models to change often and we document this well, it seems OK.

I see. I documented all this and how to invalidate the caches manually in the code and in the doc under Bicleaner -> ## Models and caching. It is a bit of a crutch but let's get back to it if we start having problems. I don't like the alternative of keeping all this logic on Takscluster since it's really hard to maintain, so sacrificing transparency here seems like the lesser of two evils.

eu9ene commented:
Ok, I think I addressed most of the issues now. Caching is not ideal but let's see how it goes. The developers might never release a new model for this tool. We can always manually invalidate caches if needed and rethink this approach if it becomes a problem.

eu9ene commented:
Ci for train-teacher:

```
[vcs 2024-01-24T18:48:08.464Z] Cloning into '/home/ubuntu/tasks/task_170612204020291/checkouts/vcs'...
[vcs 2024-01-24T18:48:20.147Z] fatal: unable to access 'https://github.com/mozilla/firefox-translations-training/': Could not resolve host: github.com
```

it looks like some temporary failure of git clone. The bicleaner steps has passed. I don't want to delay merging, so I'll rerun full CI on the next PR
